### ISSUE TITLE
Pytorch doesn't detect cuda 10-supported card

### ISSUE THREAD

**Issue by nuance** [NixOS/nixpkgs#75079]
**Describe the bug**
I have a RTX 2080 ti, which I need cuda 10.0 to use. I can see the card via nvidia-smi etc., but I can't convince pytorch to think it's available via `nix-shell`

**To Reproduce**

test_cuda.py:

```
#!/usr/bin/env nix-shell
#!nix-shell pytorch-cuda10.nix -i python3

import torch
print("Cuda: {}".format(torch.cuda.is_available()))
```

pytorch_cuda10.nix

```
with import <nixpkgs> {};

let
  unstable = import <unstable> { config.allowUnfree = true; };
in
(let
  python = let
    packageOverrides = self: super: {
      pytorch = super.pytorch.override {
        cudaSupport = true;
        cudatoolkit = unstable.cudatoolkit_10;
        cudnn = unstable.cudnn_cudatoolkit_10;
        magma = unstable.magma;
      };
    };
  in unstable.python3.override {inherit packageOverrides; self = python;};

in python.withPackages(ps: [ps.pytorch])).env
```

**Expected behavior**

I expect `./test_cuda.py` to print `Cuda: True`, but it, instead, prints `Cuda: False`

**Metadata**
Please run `nix run nixpkgs.nix-info -c nix-info -m` and paste the result.

```
$ nix run nixpkgs.nix-info -c nix-info -m
 - system: `"x86_64-linux"`
 - host os: `Linux 5.3.14, NixOS, 19.09.1476.72a2ced2523 (Loris)`
 - multi-user?: `yes`
 - sandbox: `yes`
 - version: `nix-env (Nix) 2.3`
 - channels(root): `"nixos-19.09.1498.0322870203c, unstable-20.03pre204199.3140fa89c51"`
 - channels(matt): `""`
 - nixpkgs: `/nix/var/nix/profiles/per-user/root/channels/nixos`
```

Maintainer information:
```yaml
# a list of nixpkgs attributes affected by the problem
attribute:
# a list of nixos modules affected by the problem
module:
```

**Comment by nuance** [comment]
I don't know if this is helpful, but the following is my (somewhat pared down) system-wide `configuration.nix`:

```
{ config, pkgs, lib, ... }:

let
  unstable = import <unstable> { config.allowUnfree = true; };
in
{
  imports =
    [ # Include the results of the hardware scan.
      ./hardware-configuration.nix
    ];

  nixpkgs.config.allowUnfree = true;
  boot.kernelPackages = unstable.linuxPackages_5_3;
  boot.kernelParams = [ "nordrand" ];

  # Use the systemd-boot EFI boot loader.
  boot.loader.systemd-boot.enable = true;
  boot.loader.efi.canTouchEfiVariables = true;

  hardware.enableAllFirmware = true;

  environment.systemPackages = with pkgs; [
    cudatoolkit_10
  ];

  services.xserver.videoDrivers = [ "nvidia" ];

  systemd.services.nvidia-control-devices = {
    wantedBy = [ "multi-user.target" ];
    serviceConfig.ExecStart = "${unstable.linuxPackages_5_3.nvidia_x11.bin}/bin/nvidia-smi";
  };

  # This value determines the NixOS release with which your system is to be
  # compatible, in order to avoid breaking some software such as database
  # servers. You should change this only after NixOS release notes say you
  # should.
  system.stateVersion = "19.09"; # Did you read the comment?
}
```

**Comment by nuance** [comment]
I'm not quite sure what exactly I was doing wrong, but I figured this out:

```
{ pkgs ? import <nixpkgs> {
  config.allowUnfree = true;
} }:

let
  unstable = import <unstable> { config.allowUnfree = true; };
  nvidia_x11 = unstable.linuxPackages_5_3.nvidia_x11;
  cudatoolkit = unstable.cudatoolkit_10_0;
  cudnn = unstable.cudnn_cudatoolkit_10_0;
  python =  unstable.python3.withPackages(ps: [
    ps.pytorchWithCuda
  ]);
in pkgs.stdenv.mkDerivation {
  name = "cuda-env-shell";
  buildInputs = with pkgs; [ python ];
  shellHook = ''
      export LD_LIBRARY_PATH="${nvidia_x11}/lib"
   '';
}
```

I'm not sure if this is expected behavior or a bug, or maybe missing documentation - should the LD_LIBRARY_PATH be set by default somewhere?

**Comment by andersk** [comment]
@nuance What does your system `/etc/nixos/configuration.nix` look like?

Normally PyTorch will find `libcuda.so.1` via `/run/opengl-driver/lib/libcuda.so.1`. The `/run/opengl-driver` symlink is created by `/etc/tmpfiles.d/nixos.conf` when `hardware.opengl.enable` is `true` (implied by `services.xserver.enable = true`), and should include the `libcuda.so.1` symlink if `"nvidia"` is in `services.xserver.videoDrivers`.

**Comment by nuance** [comment]
I'm not running an xserver (this is a headless box), and `hardware.opengl.enable` and `services.xserver.enable` were unset.

I added the line `hardware.opengl.enable = true;` to my configuration.nix and can now run the previous shell w/out the shellHook definition.

**Comment by nuance** [comment]
(is that expected? I'm not sure if I'm even using OpenGL when using pytorch, as I'd assume it's just using CUDA etc.?)

**Comment by andersk** [comment]
My understanding is that PyTorch with CUDA doesn’t literally use OpenGL, but it uses the same symlink mechanism configured by the `hardware.opengl.enable` option to expose `libcuda`. Perhaps the option could be given a more accurate name or an alias or better documentation.

**Comment by nuance** [comment]
The weird part to me is tensorflow seems to find the gpu / cuda without that flag set - is it a byproduct of how it’s being linked?

**Comment by andersk** [comment]
It looks like `tensorflow` pulls in `nvidia_x11` as a build input instead of locating it through the runtime symlink, which means you need to rebuild `tensorflow` for every kernel update and every new Nvidia driver release. Seems like a worse tradeoff to me.

**Comment by newkozlukov** [comment]
Hi! I'm new to Nix, is there a way to add `linuxPackages.nvidia_x11` and `LD_LIBRARY_PATH` as __runtime__ dependencies for `pytorchWithCuda`, on `nixpkgs` level?

Just ran into this problem on a headless nixos installation and first used the `shellHook`, then the `hardware.opengl.enable` - both work. However, given that `nvidia-smi` indicated no error even without `hardware.opengl.enable` option, I expected just requesting `pytorchWithCuda` in `shell.nix` to be sufficient to get things going.

Thanks

**Comment by andersk** [comment]
@newkozlukov `hardware.opengl.enable` is that way. The reason it’s not needed for `nvidia-smi` is that `nvidia-smi` is part of the `nvidia_x11` package.

To improve the UX, maybe `hardware.opengl.enable` should be auto-enabled when you add `nvidia` to `services.xserver.videoDrivers`? Right now it’s only auto-enabled by `services.xserver.enable`—but if you’re enabling the `nvidia` driver without enabling X, it’s almost certainly because you want to use CUDA.

**Comment by newkozlukov** [comment]
@andersk regarding UX, when initially configuring nixos, I felt actually hesitant to add the `services.xserver.videoDrivers` stanza. It's not immediately clear whether this would fetch Xorg server&c (which was undesired). If I may say, I'd rather have a separate option that triggers `modprobe nvidia` (and, apparently, opengl too). It would've probably simplified troubleshooting, if nixos.wiki nvidia/cuda section pointed to `hardware.opengl` as well (just noticed that tensorflow page points to `hardware.opengl.setLdLibraryPath`)

It's still not exactly the same as declaring `nvidia_x11` a "runtime dependency" (if nix has such a notion) of `pytorchWithCuda` (we set `LD_LIBRARY_PATH`/enable symlink in user's `configuration.nix`, not in `pkgs/.../pytorch/default.nix`). I skimmed the nix manual, but still amn't quite sure: if I were to override pytorch locally, would it be possible to declare __in the derivation__ that certain environment variables are to be propagated to the shell?

**Comment by andersk** [comment]
No environment variable is needed. Just configure `services.xserver.videoDrivers` and `hardware.opengl.enable`.

It’s reasonable to expect that you need some configuration: the Nvidia driver is a kernel module, and the point of Nix being purely functional is that a derivation can’t reconfigure the system just by existing.

It is confusing (but true) that the options you want for using CUDA without X and OpenGL are named with `xserver` and `opengl`; this could be improved, as could the documentation.

**Comment by newkozlukov** [comment]
Hm, I guess, in part, I'm asking because I wanted to figure out whether it's currently possible at all.

But! The `hardware.opengl.enable` creates an additional symlink which, as the shellHook option demonstrates, is not absolutely required? With the shellHook, the minimal global configuration is to modprobe the kernel module, and to add the systemd unit for control devices. So, if `pytorchWithCuda` could propagate the paths to `mkShell`, we could modify even less of global state. 

Afaiu, when nixpkgs ships _applications_, it uses `buildFHSEnv` or `patchelf` to solve dynamic linking problems. And libraries? When I say in `shell.nix` that I want to use a certain library (e.g. pytorch), what are other expected uses, if not to be able to search and dynamically link?  Do I understand it right that `mkDerivation` mechanism does not allow this? Are there some "less pure" mechanisms that would?

Thanks!

**Comment by newkozlukov** [comment]
Oh, I see, propagation of environment to reverse-dependencies in nix is implemented with setup hooks. Is that why you'd rather stick with `hardware.opengl.enable`?

**Comment by andersk** [comment]
@newkozlukov What problem do you think it would solve for the PyTorch derivation to declare an `LD_LIBRARY_PATH` environment variable?

* It would not obviate the need to edit the system configuration. Fundamentally, something needs to tell the Nvidia kernel driver to load (and which one: `nvidia`, `nvidiaBeta`, `nvidiaLegacy304`, `nvidiaLegacy340`, or `nvidiaLegacy390`).

* It would not retain the PyTorch derivation’s independence from the kernel version and Nvidia driver version that allows the extra rebuilds to be avoided. PyTorch would need to know these things in order to tell you what to set the variable to!

* It would not work outside of a build environment that runs setup hooks. For example, you would be unable to install a PyTorch using application with `nix-env`.

Fundamentally, if PyTorch knew where to look for the Nvidia library, it would just *look there*; there’d be no reason for it to *ask its dependents to set an environment variable override that forces it* to look there. But it doesn’t know, so we configure a symlink to help it—the same symlink that’s used by every OpenGL application in nixpkgs, for exactly the same reason.

The only reason anyone is even talking about the environment variable override is that it’s too easy to misconfigure the system with the symlink missing. That’s easy to fix.

**Comment by newkozlukov** [comment]
@andersk 

> What problem ... it would solve

The problem I currently see is that there's no way for `pytorch-bin` to declare (in a manner visible to the user) that during runtime it actually depends on some more system libraries, on existence of `/run/opengl-driver/lib`, and on certain kernel modules being loaded. So that when one first tries to use nix with pytorch, they might be experiencing [certain frustration](https://logs.nix.samueldr.com/nixos/2020-11-26#4289101). And it's [not specific to pytorch](https://logs.nix.samueldr.com/nixos/2020-11-26#4289160).

I'll note though, that I was confused and I'm not suggesting to modify `LD_LIBRARY_PATH` anymore. I've, since previous discussion, already realized that adding `${nvidia_x11}/lib` to `LD_LIBRARY_PATH` in a derivation's hook wouldn't make sense, because at build time we wouldn't know which `nvidia_x11` version we need. In particular, running `5.3` kernel and trying `linuxPackages_5_4.nvidia_x11`, pytorch complains about "Error 804: forward compatibility was attempted on non supported HW".

`simpson` in IRC mentions kernel capabilities: https://logs.nix.samueldr.com/nixos/2020-11-26#4289162

The way I currently see it, there's a more general use case of "declaring runtime prerequisites". A prerequisite could be a kernel module loaded or a systemd socket present.

**Comment by andersk** [comment]
The question remains: what would you expect such a declaration to do? How would the system behave differently in the presence of this declaration?

It shouldn’t stop PyTorch from building, because you might be building it on a different system than you’ll run it on. It shouldn’t stop PyTorch from installing, because the system configuration might change after installing it (insofar as “installing” is even a thing in Nix). It shouldn’t stop PyTorch from loading, because other PyTorch backends still work even if the CUDA backend isn’t available. It shouldn’t attempt to automatically load the Nvidia driver, for a number of obvious reasons (e.g. that’s a dangerous thing to allow a non-root user to do, it doesn’t know which version to use, the system might not even have an Nvidia card, or it might be using the incompatible Nouveau driver).

I think the best you can hope is that `torch.cuda.init()` raises a runtime error—which is exactly what happens now. Perhaps you might propose patching the error message with NixOS-specific advice? But I don’t think there’s any package-level declaration, even hypothetical, that would actually help anyone here.

I still think the real problem is simply the confusing names and semantics of the relevant NixOS options, and if there was one option for “enable Nvidia” or “enable CUDA” that didn’t talk about extraneous things like X and OpenGL, users wouldn’t have trouble finding it.

**Comment by newkozlukov** [comment]
@andersk

> what would you expect such a declaration to do?

Hmm... I retreat on the whole forward declaration idea... But the problem is there, and it's about "running" things. Preparing environment to run nix-built things. Outside the scope of "building" things.

- I add a pytorch-dependent derivation to the `apps` output of a flake and try `nix run .#my-app`.

  Outside nix such app would locate libcuda and GL stuff using system's impurities (e.g. I see `strace` on an opensuse machine to locate one in `/usr/lib`). With nix (whether using nixpkgs, or through mach-nix, or pip-installed in `venvShellHook`) I have modified RUNPATHs (and maybe more) and there doesn't seem to be a way to force it to find cuda (neither with `LD_LIBRARY_PATH` nor with `/run/opengl-driver`). Similarly, `nix run nixpkgs#kitty` and `nix run nixpkgs#qutebrowser` fail. Instead, I should be doing e.g. `nixGLIntel nix run nixpkgs#qutebrowser`

**Comment by NorfairKing** [comment]
I'm just finding this issue because someone pointed me to it, but have you tried setting `LD_PRELOAD="/run/opengl-driver/lib/libcuda.so"` for the python invocation. This will only work on nixos. On other systems you'll have to go find `libcuda.so`.
CC @nh2 who helped me with this.

**Comment by SomeoneSerge** [comment]
@NorfairKing indeed, `LD_PRELOAD` works, as does `nixGLNvidia` which I've been using.

This approach, however, means that on non-NixOS systems we just offload the job of runtime configuration (like solving dynamic linking issues) to the end user. This is a significant extra mental burden. In fact, Nix makes it even harder to set up the runtime environment on non-NixOS systems, as one now needs to __manually__ take care of paths to all libraries, including `glibc` and `libstdc++` to satisfy the custom linker.

This also, for example, complicates use-cases similar to using nix-provided pytorch in jupyterhub kernels (which don't inherit the environment from Nix). Each time one deals with a program similar to jupyterhub's spawners, one needs to allocate much attention to wrapping downstream programs carefully (in case of `ipykernel` - ensuring that `kernelspec` refers to the wrapped python interpreter)

This obviously goes out of scope of the github issue at hand, but I really think we need some tooling for configuration of the runtime

Thanks

**Comment by NorfairKing** [comment]
@SomeoneSerge that's fair. In practice we use a `nix-shell` with this in the shell hook:

```
      # The driver's so files need to be on the 
      if [ -e /run/opengl-driver ]; then
        # For nixos:
        export LD_PRELOAD="/run/opengl-driver/lib/libcuda.so"
      elif [ -e /etc/ld.so.cache ]; then # we are likely on NixOS-machine
        # For ubuntu
        echo "Injecting NVIDIA driver path into LD_PRELOAD using cuda_preload.py"
        export LD_PRELOAD=$(${./cuda_preload.py})
      fi
```
We have this little script to find the library on other systems:
``` python
#!/usr/bin/env python3
import subprocess
import re

libraries = subprocess.check_output(
    ["ldconfig", "-f", "/etc/ld.so.conf", "-C", "/etc/ld.so.cache", "-p"],
    universal_newlines=True,
)

# Returns `None` when the lib couldn't be found.
def find_lib(library):
    paths = []

    for match in re.finditer(
        "lib" + library + r"\.so\S* \(libc6,x86-64\) => (.+)", libraries
    ):
        paths.append(match.group(1))

    # Use the shortest path.
    if len(paths) < 1:
        return None
    return min(paths, key=len)


print(
    ":".join(
        path
        for path in [find_lib("cuda"), find_lib("nvidia-fatbinaryloader")]
        if path is not None
    )
)
```
Credit for this approach goes to @nh2

**Comment by ConnorBaker** [comment]
Closing as resolved, feel free to reopen if that's not the case!
